Add function get_variables for class calculator

peacalm opened the pull request:
Sometime we want to know what variables we should give into TokenMap before we evaluate the expression.

Change-Id: I50e1ebb70d6e1437eda8d72e412cfa73f7e020ba

VinGarcia commented on shunting-yard.cpp:
```diff
@@ -691,6 +691,16 @@ packToken calculator::eval(TokenMap vars, bool keep_refs) const {
   }
 }
 
+std::unordered_set<std::string> calculator::get_variables() const {
+  std::unordered_set<std::string> ret;
+  for (const auto& i: RPN) {
+    if (i->type == tokType::VAR) {
+      ret.insert(static_cast<Token<std::string>*>(i)->val);
+    }
+  }
+  return ret;
+}
```
Its not necessary to make these changes, but I think its better to hint at what this variable contains in its name:

```suggestion
std::unordered_set<std::string> calculator::get_variables() const {
  std::unordered_set<std::string> vars;
  for (const auto& i: RPN) {
    if (i->type == tokType::VAR) {
      vars.insert(static_cast<Token<std::string>*>(i)->val);
    }
  }
  return vars;
}
```

VinGarcia commented on shunting-yard.h:
```diff
@@ -88,7 +90,19 @@ struct TokenUnary : public TokenBase {
 };
 
 class packToken;
-typedef std::queue<TokenBase*> TokenQueue_t;
+
+// Adapt to std::queue<TokenBase*>
+class TokenQueue_t: public std::deque<TokenBase*> {
+public:
+  void push(TokenBase* t) {
+    push_back(t);
+  }
+  void pop() {
+    pop_front();
+  }
+};
+
+
```
I don't think this change is necessary, do you have an argument in favor of keeping this? Since this is not even being used I am not sure we should keep it.

VinGarcia replied:
Oh I haven't noticed the type changed xP, that's why I did this comment. My fault here.

VinGarcia commented on test-shunting-yard.cpp:
```diff
@@ -1173,3 +1173,11 @@ TEST_CASE("Exception management") {
   REQUIRE_THROWS(ecalc2.compile("map(['hello']]"));
 }
+
+TEST_CASE("Get variables") {
+  calculator c("a + sin(b) - c**2 / d");
+  auto tmp = c.get_variables();
+  auto vars = std::set<std::string>(tmp.begin(), tmp.end());
+  auto expected = std::set<std::string>{"a", "b", "c", "d"};
+  REQUIRE(vars == expected);
+}
```
Using inline expressions is useful in a few ways, one of them is that it saves you the trouble of coming up with names for variables that will only be used once. In this case, the name `tmp` is not informative so we might as well not give it a name, e.g.:

```suggestion
TEST_CASE("Get variables") {
  calculator c("a + sin(b) - c**2 / d");
  auto expectedVars = std::unordered_set<std::string>{"a", "b", "c", "d"};
  REQUIRE(c.get_variables() == expectedVars);
}
```

VinGarcia reviewed (changes requested):
Hey @peacalm thanks for this improvement, looks good I just need to understand if we need to add that `push()` and `pop()` functions on the `TokenQueue_t`.

I also added some suggestions tell me what you think of them too =]

peacalm commented:
> Hey @peacalm thanks for this improvement, looks good I just need to understand if we need to add that `push()` and `pop()` functions on the `TokenQueue_t`.
> 
> I also added some suggestions tell me what you think of them too =]

😄, your suggestions are great, I accepted them all. 

I add `push()` and `pop()` for TokenQueue_t just want to make it compatible with existing code, because I changed TokenQueue_t from type `std::queue` to `std::deque` to make items in the queue visitable. 
Or another way: just modify `typedef std::queue<TokenBase*> TokenQueue_t;` to `typedef std::deque<TokenBase*> TokenQueue_t;`, and modify every `push()` to `push_back()`, every `pop()` to `pop_front()`.

All I want is a feature to know what variables are required in some given expression, so I must search them in TokenQueue, but items in `std::queue` are not iterable.

My code looks like a short hack to support what I want, maybe you have more consideration about whether to support this feature, or maybe you have better implementation. Never mind, it's just a little idea after I reading this project. Thanks very much for your feedback.

VinGarcia commented:
@peacalm no it's a great addition, yeah I will think about the implementation details, I will try reading these changes more carefully tonight and thinking about whether there is anything I should consider about this type change. 

Thanks for contributing.

VinGarcia reviewed (approved):
Hey @peacalm the implementation is great, and this change on the type might have a minor performance issue, but I can test that and improve on it later on, I think we can merge this one =]

VinGarcia commented:
Thanks for this contribution =]

VinGarcia commented:
You are welcome @peacalm, thank you actually.

Yeah, so I think that for actually choosing which tool to use between `std::queue` and `std::deque` we should actually write some benchmark. I also believe this probably behaves the same as before so this is probably not a real issue. If we wanted to really tackle this problem I would start by writing a benchmark, but I will probably not have time for that =P.

And I am happy it's been helpful to you =].
